src/concepts/imperative.md

```diff
@@ -0,0 +1,167 @@
+types. Technically, it's an infix function that takes in the `'a ref` to
+update and the `'a` value to update it to.
+
+```sml
+(* := : 'a ref * 'a -> unit *)
+val pointer : int ref = ref 7
+val x : int = !pointer
+val () : unit = pointer := 21
+val y : int = !pointer
+```
+
+Here, the value `7 : int` would be bound to `x`. However, the value `21 : int`
+would be bound to `y`. When we dereference our ref-cell the first time, the
+value is whatever we initialized it to (`7`). However, after we change what's
+inside the ref-cell, we dereference the updated value (`21`). Even though
+`x` and `y` are both bound to `!pointer`, they have different values!
+
+> 💡 Comparison to the `void`!
+>
+> Remember that SML is a functional programming language. This means that
+> anything that's not an exception or infinite loop must be a value! As
+> such, when we do things with side effects (like assignment), we can't
+> really have _no value_ being returned. As such, SML implements this by
+> having these side effect functions and operations return `() : unit`.
+> Theoretically, `:=` could return something crazy, like an `'a tree list option`
+> if we really wanted to. It's just that `() : unit` is convenient.
+> So when you see `:= : 'a * 'a ref -> unit`, you can think of it as a
+> `void` method or function in Java, C, and C++.
+>
+> In other words, returning `() : unit` for the assignment operator is
+> SML's way of implementing `void` methods.
+
+Introducing imperative logic causes difficulties like these. It takes away the
+confidence we have in our values instilled in us by referential transparency.
+It often makes doing formal proofs about imperative code a bit more difficult.
+
+### Sequential Evaluation `;`
+
+Finally, having a way to execute side effects but still return something
+meaningful will be useful to use. Generally, `;` (yes this is just a
+semicolon lol) lets us "execute" two "programs" in sequential order.
+You see this used in the REPL:
+
+```
+val x = 2 + 2;
+val y = 8 - 4;
+```
+
+Where we execute one "program" with the side effect of binding `2 + 2`
+to `x`, and then execute the next "program" with the side effect of
+binding `8 - 4` to `y`, and continue to wait for the next "program".
+If you're familiar with languages like Java and C, this can also be
+used to explain why semicolons are required at the end of each line.
+
+Where the semicolon `;` is used to mark the end of one program and set up
+the next program to sequentially follow. In SML, this is also valid syntax.
+
+> For some expressions `exp1 : t1`, `exp2 : t2`, we have that `exp1;exp2 : t2`
+> where `exp1` is evaluated first, and then `exp2` evaluated thereafter.
+
+One thing to note is that in SML, only the value of the second expression
+is returned, and the type of the whole expression is the type of the last
+expression.
+
+```sml
+val pointer : int ref = ref 0
+val x : int = pointer := 7; !pointer
+val y : int = pointer := 21; !pointer
```

Comment:
```suggestion
val x : int = (pointer := 7; !pointer)
val y : int = (pointer := 21; !pointer)
```
You more than likely mean this instead---without parens the semicolon indicates two sequential _programs_ not expressions. If you type `val x = 7 ; 8` in the REPL you can probably see what I mean.

Reply:
Ah didn't know that, thanks